### ISSUE TITLE
use-package :if keyword does not work with straight.el

### ISSUE THREAD

**Issue by caadar** [raxod502/straight.el#132]
According to the manual one of the features is support for conditional installation, as well as use-package integration.

But :if keyword of use-package does not work with Straight.

E.g.:

>     (use-package bm
>       :if (not window-system))

unconditionally always loads bm.

So, what the proper way to do conditional installation? How to conditionally disable package install, may be by means of :files' directive?

**Comment by raxod502** [comment]
So there's a couple of things here.

Firstly, what I meant by conditional installation was that it's possible to tell `straight.el` about a package without immediately installing and loading it. That's actually separate from the `:if` keyword of `use-package`. And I should probably be clear about this in the documentation.

Secondly, it appears that the `:if` keyword does not work when using `straight.el` with `use-package`. That's a bug. In other words, what you called conditional installation was different from what I called conditional installation, but both should be supported.

Thirdly, it's on the roadmap to merge these two previous points: that is, special-case the `:if` directive so that it does the only-register-but-don't-install mechanism when `:if` is disabled, and does the regular register-and-install mechanism otherwise. This feature is tracked by https://github.com/raxod502/straight.el/issues/45. 

So in summary:

- this is an unexpected bug
- the documentation should also be fixed

The `:files` directive doesn't have anything to do with it.

**Comment by raxod502** [comment]
Cursory inspection yields the following insight:

```
(macroexpand-1 '(use-package foo :if nil))

(progn
  (straight-use-package-ensure-function
   (quote foo) (quote t) (quote (:deferred t)) :ensure)
  (when nil))
```

That is, the `:ensure` function is called outside of the `when` clause.

I have a question for you: are you sure that `:if` correctly prevents `package.el` from installing a package as well? In other words, are you sure that this is a problem with `straight.el`, rather than a problem with `use-package`?

**Comment by raxod502** [comment]
(Of course, it might be my fault after all: I was the one who added generalized `:ensure` support to `use-package` in the first place!)

**Comment by caadar** [comment]
Before report I check bare use-package. It works as expected for me.

**Comment by caadar** [comment]
In the meantime obvious workaround may be used:

>     (unless window-system
>       (use-package bm))

**Comment by raxod502** [comment]
I just tested and AFAICT this is not a `straight.el` problem; it also occurs when using `use-package` with `package.el`. You should probably open an issue against `use-package` instead.

**Comment by caadar** [comment]
Let's test it again.

(Always run Emacs as \`emacs -Q'.)

1. Prepare.

>     (require 'package)
>     (package-initialize)
>     (add-to-list 'package-archives '("melpa" . "http://melpa.milkbox.net/packages/") t)
>     (package-refresh-contents)
>     (package-install 'use-package)

2. Restart Emacs.

Evaluate:

>     (require 'package)
>     (package-initialize)
>     (require 'use-package)
>     window-system => x
>     (use-package async
>         :ensure t
>         :if (not window-system))
>     (featurep 'async) => nil

3. Restart Emacs.

Evaluate:

>     (require 'package)
>     (package-initialize)
>     (require 'use-package)
>     window-system => x
>     (use-package async
>         :ensure t
>         :if window-system)
>     (featurep 'async) => t

So, use-package works as expected with package.el.

I also test it without package.el loaded, simple by

>     (add-to-list 'load-path "~/.emacs.d/straight/build/use-package")
>     (add-to-list 'load-path "~/.emacs.d/straight/build/bind-key")
>     (add-to-list 'load-path "~/.emacs.d/straight/build/async")
>     (require 'use-package)
>     .....
Am I missing something?

**Comment by caadar** [comment]
...but if async is not installed, use-package installes it unconditionally.

**Comment by raxod502** [comment]
That's what I thought. Indeed, I only tested about whether `async` was installed conditionally, not whether it was loaded conditionally.

You're saying that the loading does happen conditionally, even though the installation happens unconditionally, for `package.el`? (I haven't run your test yet, but will later.)

**Comment by caadar** [comment]
Just so.

**Comment by raxod502** [comment]
OK. I understand what you are saying now. The confusion is that there are two different concerns:

1. whether the package is made available by the package manager (i.e. autoload evaluation and `load-path` extension) in the case of `:if nil`
2. whether the feature is actually required in the case of `:if nil`

I did not even think of case (1) since I turned on `use-package-always-defer` months ago. But now I see that case (1) was what you meant when you said "loading a package".

This means that your results in https://github.com/raxod502/straight.el/issues/132#issuecomment-322576647 (i.e. that `package.el` works correctly w.r.t case (1)) make perfect sense to me. However, I tested your original issue, and I can't reproduce it. That is, after I evaluate

```
(use-package bm
  :if (not window-system))
```

then `(featurep 'bm)` returns `nil` if I'm in windowed Emacs. Both when using `package.el` and when I'm using `straight.el`.

For case (2), there's a bug in `use-package`, so case (2) does not work in `package.el` nor `straight.el`.

**Comment by raxod502** [comment]
Let me know if you can find a recipe to reproduce the original bug report from `emacs -Q`.

**Comment by raxod502** [comment]
> ...but if async is not installed, use-package installes and loads it unconditionally. (Why?!)

Just in case I didn't make this clear, this behavior is because of the bug on `use-package` vis-a-vis case (2) in https://github.com/raxod502/straight.el/issues/132#issuecomment-322632626.

Although, perhaps the upstream doesn't consider it a bug, but rather desired behavior. I'm not going to ping @ jwiegley since he's stated multiple times that he doesn't use the `:ensure` keyword in `use-package` and can't maintain it. Not sure who to bring into the discussion, if anyone.

**Comment by caadar** [comment]
I can't reproduce it with `emacs -Q`. Surprisedly I can't reproduce it with working config anymore. Works just fine, except case (2) oddities though.

Feel free to close the issue after documentation correction, if needed.

**Comment by raxod502** [comment]
I'm not actually sure what I could change in the docs to make this clearer. I'll most likely leave things as is for the moment, but if somebody else is confused (or has suggestions on how to improve), I'll update it.

**Comment by caadar** [comment]
[github.com/jwiegley/use-package/issues/496](https://github.com/jwiegley/use-package/issues/496)
